tests/torch_trainer.py

```diff
@@ -0,0 +1,141 @@
+        loss_func = lambda *args, **kwargs: torch.nn.functional.mse_loss(*args, **kwargs)
+
+        # Define the training loop
+        def sgd_train_func(module_func, loss_func, optim_func, params, bufs, optim_state, batch):
+            inputs = batch[0]
+            targets = batch[1]
+
+            # wrap forward pass + loss computation in a function
+            def compute_loss(params, bufs, inputs, targets):
+                # do forward pass
+                bufs, out = module_func(params, bufs, inputs)
+
+                # do some debugging when in local mode
+                if atorch.mode() == "local":
+                    print("out: ", out)
+
+                # do loss computation
+                loss_value = loss_func(out, targets)
+                return loss_value, bufs
+
+            if atorch.mode() == "dist":
+                compute_loss = alpa.automatic_layer_construction(layer_num=2)(compute_loss)
+
+            # do model forward + backward pass
+            (loss_value, bufs), params_grad = atorch.value_and_grad(compute_loss, has_aux=True)(params, bufs, inputs, targets)
+
+            # do optimizer step
+            params, optim_state = optim_func(params, params_grad, optim_state)
+
+            return params, bufs, optim_state, loss_value
+
+        train_func = functools.partial(sgd_train_func, module_func, loss_func, optim_func)
+
+        if atorch.mode() == "dist":
+            parallel_method = alpa.PipeshardParallel(stage_mode="auto", num_micro_batches=num_micro_batches)
+            # parallel_method = alpa.PipeshardParallel(stage_mode="auto", num_micro_batches=num_micro_batches, submesh_physical_shape_space="all", submesh_logical_shape_space="all")
+            # parallel_method = alpa.PipeshardParallel(stage_mode="uniform", num_micro_batches=num_micro_batches)
+            # parallel_method = alpa.ShardParallel(num_micro_batches=num_micro_batches)
+            # parallel_method = alpa.ShardParallel()        
+            train_func = alpa.parallelize(
+                train_func,
+                method=parallel_method,
+                batch_argnums=(3,),  # NOTE: assumes the 4th argument is input batch
+                donate_argnums=(0, 1, 2),  # NOTE: preserves memory addr and sharding spec for first 3 args
+            )
+
+        def iter_func(params, bufs, optim_state, pt_batch):
+            params, bufs, optim_state, loss_value = train_func(
+                params,
+                bufs,
+                optim_state,
+                atorch.to_format(atorch.mode(), pt_batch),
+            )
+            # Show that outputs are in Alpa format (no need in actual training code)
+            atorch.assert_format(atorch.mode(), params, bufs, optim_state, loss_value)
+            return params, bufs, optim_state, loss_value
+
+        # Bind the torch operators to Alpa implementation when in dist mode.
+        with atorch.bind_ops(atorch.mode() == "dist"):
```

Comment:
can the `bind_ops` context be automatically incorporated when `mode == dist`? This API seems to me a bit extra

Reply:
move bind_ops into generated python source

Reply:
weight_init_func should use atorch